core/civilization/person/ear/default.py

```diff
@@ -0,0 +1,123 @@
+            t.start()
+
+    def bind_port(self):
+        port_start = settings["PORT_START"]
+        port_range = settings["PORT_RANGE"]
+        host = settings["HOST"]
+        for port in range(port_start, port_start + port_range):
+            try:
+                self.listener_socket.bind((host, port))
+                self.port = port
+                return True
+            except socket_error:
+                pass
+        return False
+
+    def get_instruction_length(self, conn: socket) -> int:
+        instruction_length_data = conn.recv(INSTRUCTION_LENGTH_BYTES)
+        return unpack(">I", instruction_length_data)[0]
+
+    def get_extra_length(self, conn: socket) -> int:
+        extra_length_data = conn.recv(EXTRA_LENGTH_BYTES)
+        return unpack(">I", extra_length_data)[0]
+
+    def get_sender(self, conn: socket) -> BasePerson:
+        message_sender_data = conn.recv(MESSAGE_SENTER_BYTES)
+        message_sender = sub("[\0]", "", message_sender_data.decode())
+        if message_sender not in self.person.friends:
+            raise Exception(INVALID_MESSAGE_SENDER_ERROR_MESSAGE)
+        return self.person.friends[message_sender]
+
+    def get_message_type(self, conn: socket) -> MessageType:
+        message_type_data = conn.recv(MESSAGE_TYPE_BYTES)
+        message_type = (unpack(">I", message_type_data))[0]
+        if message_type not in [e.value for e in MessageType]:
+            raise Exception(INVALID_MESSAGE_TYPE_ERROR_MESSAGE)
+        return MessageType(message_type)
+
+    def get_message(self, conn: socket, message_length: int) -> str:
+        message_body_data = conn.recv(message_length)
+        return message_body_data.decode()
+
+    def parse_data(self, conn: socket) -> Tuple[BasePerson, MessageType, str, str]:
+        instruction_length = self.get_instruction_length(conn)
+        extra_length = self.get_extra_length(conn)
+        sender = self.get_sender(conn)
+        message_type = self.get_message_type(conn)
+        instruction = self.get_message(conn, instruction_length)
+        extra = self.get_message(conn, extra_length)
+        return (sender, message_type, instruction, extra)
+
+    def listen(self):
+        self.listener_socket.listen()
+        print(f"[{self.person.name}-Ear] : Listening On {self.port}")
+        while True:
+            conn, addr = self.listener_socket.accept()
+            if addr[0] != settings["HOST"]:
+                print(
+                    System.announcement(f"[{self.person.name}-Ear] : Unknown Message")
+                )
+                continue
+            try:
+                message_sender, _, instruction, extra = self.parse_data(conn)
+            except Exception as e:
+                print(
+                    System.announcement(
+                        f"[{self.person.name}-Ear] : Parsing Message Failed\n{e}"
+                    )
+                )
+                continue
+            self.person.respond(
```

Comment:
How about handing this method over to callback?

Reply:
you mean `def listen(self, callback)`?

Reply:
yes exactly

Reply:
It doesn't matter but is there any other places that can call `listen`? For now message format is only for `talk` and `respond`, as it contains `instruction`, `extra`. I think using callback means `Person` can listen differently to the message.

Reply:
it's ambiguous. let's skip now.